### ISSUE TITLE
`map!` macro should not silently ignore some settings

### ISSUE THREAD

**Issue by dschrempf** [hlissner/doom-emacs#5532]
### Describe your request

Thank you for providing this Emacs configuration. It has been a really pleasent experience, so far! I have a suggestion regarding the `map!` macro that would have saved me some time.

Two observations:
  - `map!` silently ignores `:map` when `:leader` is present.
  - `map!` silently ignores `:desc` when `[remap a] #'b` is used.
 
The second issue is not a problem at all, more of an annoyance that I cannot change the description when remapping functions in specific major modes.

The first issue was a problem to me, because first, I tried to achieve major-mode specific bindings involving the `<leader>` key like so:
```
  (map! :leader :desc "Cite insert" "n i" #'bibtex-actions-insert-citation)
  (map! :leader :desc "Org cite insert" :map org-mode-map "n i" #'org-cite-insert)
```
Which didn't work, because the second map is just overwriting the first one. After some debugging, I found https://github.com/hlissner/doom-emacs/issues/3978 and https://github.com/hlissner/doom-emacs/issues/2718. Then it was clear to me, that I have to use `remap`. However, it was not clear from the help text of `map!`.

I suggest improving the help message of `map!` and that `map!` should fail if arguments are ignored.

Maybe other options are also ignored, I don't know.

### Briefly explain its use-case

The use case is that debugging time is decreased.

Thanks for your opinion on this!

**Comment by hlissner** [comment]
I agree with this in spirit, but I am currently swamped, so it'll be some time before I address it. 

The `:leader`+`:map` issue is documented in `map!`'s docstring and the `:desc` issue is a limitation of `general` and `which-key`, rather than `map!` (after all, `map!` is really just a thick wrapper around the two). You can bind a key with general, with a description, but if that key is a remap, it gets silently ignores. I'm not convinced it should be `map!`'s responsibility to handle the error in this case.

That said, I do plan to rewrite `map!` to be more robust and factor out its dependency on `general`, at which point I'll consider better error handling, but it's very low priority and is a non-trivial undertaking for disproportionately little gain.

In any case, I'm open to PRs if someone wants to have a crack at it. I'll let know when there is movement on this.

**Comment by dschrempf** [comment]
I understand!

I was checking the help text of `map!` and didn't see where the `:leader :map` issue is explained. I quote:
```
  :leader [...]                   an alias for (:prefix doom-leader-key ...)
  :prefix [PREFIX] [...]          set keybind prefix for following keys. PREFIX
                                  can be a cons cell: (PREFIX . DESCRIPTION)
```
It was not clear to me that the leader prefix is connected to the global key map. In the end, I could just do:
```
  (map! :desc "Org cite insert" :map org-mode-map "SPC n i" #'org-cite-insert)
```
and it would work only in `org-mode-map`. I could work on the documentation, if you like, but I don't see how I could understand and change the implementation of `doom--map-process`.

**Comment by lelit** [comment]
I do not know if my case fall into this issue, but I'm having trouble adding new bind to an existing keymap, where the `:desc` isn't honored.
I tried the following:
```elisp
(map!
 :map doom-leader-toggle-map
 :desc "Menu bar" "m" #'menu-bar-mode)
```
but when the `C-t` popup appears, the `m` entry shows "menu-bar-mode", not "Menu bar".

In the end, I added an explicit
```elisp
(which-key-add-key-based-replacements "C-c t m" "Menu bar")
```
that apparently workaround the issue.
What am I missing?

**Comment by kavasaziz** [comment]
Just for the sake of record, when I added `(doom! :config (default +bindings))` into `init.el`, `desc:` fields started to work.
